### PR TITLE
Align argument types and format strings

### LINKED ISSUE #523: Create agreement between argument type and format string
> Please correct before 2023-12-10 to safely retain your package on CRAN.

The relevant warning can be seen in GHA:

```
── R CMD check ─────────────────────────────────────────────────────────────────
* using log directory ‘/home/runner/work/vroom/vroom/check/vroom.Rcheck’
* using R Under development (unstable) (2023-11-26 r8[56](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:58)38)
* using platform: x86_64-pc-linux-gnu
* R was compiled by
    gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
    GNU Fortran (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
* running under: Ubuntu 22.04.3 LTS
* using session charset: UTF-8
* using options ‘--no-manual --as-cran’
* checking for file ‘vroom/DESCRIPTION’ ... OK
* this is package ‘vroom’ version ‘1.6.4.9000’
* package encoding: UTF-8
* checking package namespace information ... OK
* checking package dependencies ... OK
* checking if this is a source package ... OK
* checking if there is a namespace ... OK
* checking for executable files ... OK
* checking for hidden files and directories ... OK
* checking for portable file names ... OK
* checking for sufficient/correct file permissions ... OK
* checking whether package ‘vroom’ can be installed ... [61s/[59](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:61)s] WARNING
Found the following significant warnings:
  vroom_big_int.h:59:30: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_chr.h:41:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_dttm.h:[68](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:70):27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_date.h:50:27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_dbl.h:40:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_fct.h:125:29: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_int.h:38:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_num.h:50:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_time.h:50:27: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_rle.h:43:26: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘R_xlen_t’ {aka ‘long int’} [-Wformat=]
  vroom_write.cc:2[73](https://github.com/tidyverse/vroom/actions/runs/7021455235/job/19103553319#step:7:75):16: warning: field precision specifier ‘.*’ expects argument of type ‘int’, but argument 2 has type ‘std::vector<char>::size_type’ {aka ‘long unsigned int’} [-Wformat=]
See ‘/home/runner/work/vroom/vroom/check/vroom.Rcheck/00install.out’ for details.
* used C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
* used C++ compiler: ‘g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
```

### PR DESCRIPTION
**Pull request by jennybc**
Closes #523

### CODE REVIEW

File: src/utils.h [hunk 1]
```diff
@@ -5,6 +5,7 @@
 #include <cstring>
 #include <sstream>
 #include <string>
+#include <tuple>
```
Comment (jennybc) [inline src/utils.h:8]: I can't say that I completely understand why the need for this was surfaced by the changes in the content and usage of `r_utils.h`. But I'm also not sure that I really need to get to the bottom of it.

File: src/r_utils.h [hunk 2]
```diff
@@ -9,6 +9,14 @@
...
 #include <vector>
 
+#ifndef R_PRIdXLEN_T
+#  ifdef LONG_VECTOR_SUPPORT
+#    define R_PRIdXLEN_T "td"
+#  else
+#    define R_PRIdXLEN_T "d"
+#  endif
+#endif
+
```
Comment (jennybc) [inline src/r_utils.h:19]: Inspired by https://github.com/wch/r-source/blob/75cd3117e23f2010634200875f2ef50a83461b8f/src/include/Rinternals.h#L71C1-L80

Who/what is going to define `LONG_VECTOR_SUPPORT` 🤔 ?

Comment (jennybc) [reply src/r_utils.h:19]: I suppose it comes via `cpp11/R.hpp`, which includes `Rinternals.h`.

Comment (gaborcsardi) [reply src/r_utils.h:19]: Yeah, it is coming from the R headers, you are allowed to use it.

**Review by gaborcsardi — APPROVED** [review]
This looks pretty good to me, much better than casting. But just to make sure, would you mind adding these to the CI matrix, temporarily?
- windows 4.2
- windows R-devel

Windows 4.2, because it uses yet another compiler that we do not test for currently, and R-devel because that's where the recent changes are.

I am still not entirely sure if it is going to work without a warning on older Linux, e.g. Centos/RedHat 7, I'll check it soon. 

EDIT: seems fine on CentOS 7 as well (with gcc 4.8.5!), with R-release and R-devel.

**Comment by jennybc** [comment]
Checking on Windows with r-devel was revealing. The existing fixes all hold up, but I get some new hits in a totally different file, so I'm happy to learn about this now! I also sent everything to win-builder (devel, release, oldrel) and get completely consistent results: the new warnings seen below on devel, clean pass on release and oldrel.

```
Found the following significant warnings:
  unicode_fopen.h:52:52: warning: conversion lacks type at end of format [-Wformat=]
  unicode_fopen.h:52:14: warning: too many arguments for format [-Wformat-extra-args]
  unicode_fopen.h:78:52: warning: conversion lacks type at end of format [-Wformat=]
  unicode_fopen.h:78:14: warning: too many arguments for format [-Wformat-extra-args]
```

**Pull request merged**